// @openai/apps-sdk-ui/dist/es/lib/dateUtils.js
import { DateTime } from "luxon";
/**
 * Get all days in a given month as DateTime objects
 */
export const getDaysOfMonth = (date) => {
    const startOfMonth = date.startOf("month");
    const daysInMonth = startOfMonth.daysInMonth ?? 31; // fallback to 31 days if undefined
    const days = [];
    for (let day = 0; day < daysInMonth; day++) {
        const dayDate = startOfMonth.plus({ days: day }).startOf("day");
        days.push(dayDate);
    }
    return days;
};
/**
 * Check if the first date is before the second date
 */
export const isBefore = (targetDate, compareDate) => targetDate.toMillis() < compareDate?.toMillis();
/**
 * Check if the first date is after the second date
 */
export const isAfter = (targetDate, compareDate) => targetDate.toMillis() > compareDate?.toMillis();
/**
 * Check if two dates are the same day (ignoring time)
 */
export const isSameDay = (targetDate, compareDate) => targetDate.hasSame(compareDate, "day");
/**
 * Check if a date is today
 */
export const isToday = (dateTime) => isSameDay(dateTime, DateTime.local());
/**
 * Split an array of days into weeks (arrays of 7 days each)
 */
export const chunkIntoWeeks = (daysInMonth, size = 7) => {
    const weeks = [];
    for (let i = 0; i < daysInMonth.length; i += size) {
        weeks.push(daysInMonth.slice(i, i + size));
    }
    return weeks;
};
/**
 * Get the start and end date of the month the provided date is in, with optional min and max limits.
 */
export const getMonthStartAndEnd = (date, { min, max } = {}) => {
    const startOfMonth = date.startOf("month");
    const endOfMonth = date.endOf("month");
    return [
        min ? DateTime.max(min, startOfMonth) : startOfMonth,
        max ? DateTime.min(max, endOfMonth) : endOfMonth,
    ];
};
